vsintegration/src/FSharp.Editor/CodeFix/ConvertToAnonymousRecord.fs

```diff
@@ -3,45 +3,49 @@
 namespace Microsoft.VisualStudio.FSharp.Editor
 
 open System.Composition
-open System.Threading
 open System.Threading.Tasks
 open System.Collections.Immutable
 
 open Microsoft.CodeAnalysis.Text
 open Microsoft.CodeAnalysis.CodeFixes
-open Microsoft.CodeAnalysis.CodeActions
+
+open CancellableTasks
 
 [<ExportCodeFixProvider(FSharpConstants.FSharpLanguageName, Name = CodeFix.ConvertToAnonymousRecord); Shared>]
 type internal FSharpConvertToAnonymousRecordCodeFixProvider [<ImportingConstructor>] () =
     inherit CodeFixProvider()
 
-    static let title = SR.ConvertToAnonymousRecord()
-
     override _.FixableDiagnosticIds = ImmutableArray.Create("FS0039")
 
-    override _.RegisterCodeFixesAsync context : Task =
-        asyncMaybe {
-            let document = context.Document
+    interface IFSharpCodeFix with
+        member _.GetChangesAsync document span =
+            cancellableTask {
+                let! cancellationToken = CancellableTask.getCurrentCancellationToken ()
 
-            let! parseResults =
-                document.GetFSharpParseResultsAsync(nameof (FSharpConvertToAnonymousRecordCodeFixProvider))
-                |> liftAsync
+                let! parseResults = document.GetFSharpParseResultsAsync(nameof (FSharpConvertToAnonymousRecordCodeFixProvider))
 
-            let! sourceText = context.Document.GetTextAsync(context.CancellationToken)
+                let! sourceText = document.GetTextAsync(cancellationToken)
 
-            let errorRange =
-                RoslynHelpers.TextSpanToFSharpRange(document.FilePath, context.Span, sourceText)
+                let errorRange =
+                    RoslynHelpers.TextSpanToFSharpRange(document.FilePath, span, sourceText)
 
-            let! recordRange = parseResults.TryRangeOfRecordExpressionContainingPos errorRange.Start
-            let! recordSpan = RoslynHelpers.TryFSharpRangeToTextSpan(sourceText, recordRange)
+                let changes =
+                    parseResults.TryRangeOfRecordExpressionContainingPos errorRange.Start
+                    |> Option.bind (fun range -> RoslynHelpers.TryFSharpRangeToTextSpan(sourceText, range))
+                    |> Option.map (fun span ->
+                        [
+                            TextChange(TextSpan(span.Start + 1, 0), "|")
+                            TextChange(TextSpan(span.End - 1, 0), "|")
```

Comment:
Good  catch.

So the old-old code was actually quite risky with the chained .WithChanges code.
We should definitely stick with passing a collection of TextChanges, since Roslyn knows how to process those correctly in a single atomic change.

Reply:
Yep. We have this in a couple other code fixes now, will address them in the followup.